### ISSUE TITLE
[FEATURE] Refactor TransformationTracker once again

### ISSUE THREAD

**Issue by konnov** [informalsystems/apalache#534]
I am adding everbody on this issue, as it deserves discussion.

We are using transformation trackers to make sure that the modified code is properly backtracked to the source code. Currently, `TransformationTracker.trackEx` is declared like this:

```scala
  type TlaExTransformation = TlaEx => TlaEx
 ...
  def trackEx(tr: TlaExTransformation): TlaExTransformation
```

It has become a pattern in all of our transformers to write a check similar to the following one in `ReplacedFixed`:

```scala
        val retEx = if ( defs == newDefs && body == newBody ) ex else LetInEx( newBody, newDefs : _* )
```

The above check is annoying and it introduces unnecessary complexity. Moreover, the tracker itself is comparing identifiers of the old and new expressions. I propose to introduce two case classes, similar to `Either` or `Option`:

```scala
abstract class TransformationResult
case class Changed(newEx: TlaEx) extends TransformationResult
case class Unchanged(oldEx: TlaEx) extends TransformationResult
```

The reason for not using `Either` is that it is hard to remember what is `Left` and what is `Right`.

Then we can refactor `trackEx` as follows:

```scala
  def trackEx(tr: TlaEx => TransformationResult[TlaEx]): TlaEx => TransformationResult[TlaEx]
```

So the user of this class has to explicitly tell the tracker, whether the expression has been transformed or not. The tracker can do an additional consistency check, to make sure that the user has not passed a new expression in `Unchanged` (by comparing IDs, which is inexpensive).

If we introduce nice join methods to `TransformationResult`, we can propagate unchanged expressions upwards the recursion. This will make the above coding pattern explicit and protect us from making mistakes in the transformations.

**Comment by shonfeder** [comment]
Won't using `Either` give us all the standard library methods on the type for free? Won't that be the worth the cost of remembering which side is which?

iiuc, this might be a good fit for the Writer monad: http://www.smartjava.org/content/scalaz-features-everyday-usage-part-3-state-monad-writer-monad-and-lenses/#writer-monad Is worth considering a refactor to use a Writer monad for logging the transitions that would let us develop a more regular structure more susceptible to principled manipulation?

**Comment by Kukovec** [comment]
I don't think this would add much. Now, instead of the above check, you'd have to do:
```
if (newDefs.forall { _.isInstanceOf[Unchanged] }) Unhanged(ex)
else Changed(...)
```

**Comment by konnov** [comment]
we can define nice join operators like in `Either`

**Comment by shonfeder** [comment]
Or just use `Either` and not have to define anything, and it'd also be clear what the semantics are to new readers :)

**Comment by konnov** [comment]
We should probably think about having some flow framework between transformation results, like people to in static analysis. Many bugs come from the fact that different preprocessing steps should trigger other preprocessing steps when something has changed.
